plugins/regex_revalidate/regex_revalidate.cc

```diff
@@ -0,0 +1,729 @@
+      if (rules_changed) {
+        pstate->min_expiry = 0;
+      }
+    }
+  }
+
+  // prune
+  if (pstate->min_expiry < timenow) {
+    if (newrules.empty()) {
+      newrules = *(pstate->invalidate_vec);
+    }
+
+    time_t new_expiry  = 0;
+    bool const pruned  = pruneConfig(timenow, &newrules, &new_expiry);
+    pstate->min_expiry = new_expiry;
+
+    if (pruned) {
+      rules_changed = true;
+    }
+  }
+
+  if (rules_changed) {
+    auto vecnew = std::make_shared<std::vector<Invalidate>>(std::move(newrules));
+
+    std::atomic_store(&(pstate->invalidate_vec), vecnew);
+    DEBUG_LOG("New configuation installed");
+    pstate->logConfig();
+  }
+
+  // Reschedule the continuous pruning/reload job
+  if (TS_EVENT_TIMEOUT == event) {
+    TSContScheduleOnPool(config_cont, CONFIG_TMOUT, TS_THREAD_POOL_TASK);
+  }
+
+  return 0;
+}
+
+time_t
+get_date_from_cached_hdr(TSHttpTxn txn)
+{
+  TSMBuffer buf;
+  TSMLoc hdr_loc, date_loc;
+  time_t date = 0;
+
+  if (TSHttpTxnCachedRespGet(txn, &buf, &hdr_loc) == TS_SUCCESS) {
+    date_loc = TSMimeHdrFieldFind(buf, hdr_loc, TS_MIME_FIELD_DATE, TS_MIME_LEN_DATE);
+    if (nullptr != date_loc) {
+      date = TSMimeHdrFieldValueDateGet(buf, hdr_loc, date_loc);
+      TSHandleMLocRelease(buf, hdr_loc, date_loc);
+    }
+    TSHandleMLocRelease(buf, nullptr, hdr_loc);
+  }
+
+  return date;
+}
+
+int
+main_handler(TSCont cont, TSEvent event, void *edata)
+{
+  TSHttpTxn txn = (TSHttpTxn)edata;
+  int status    = TS_ERROR;
+
+  switch (event) {
+  case TS_EVENT_HTTP_CACHE_LOOKUP_COMPLETE:
+    if (TSHttpTxnCacheLookupStatusGet(txn, &status) == TS_SUCCESS) {
+      if (status == TS_CACHE_LOOKUP_HIT_FRESH) {
+        PluginState *const pstate = static_cast<PluginState *>(TSContDataGet(cont));
+        // atomcally grab a handle
+        using InvHandle        = std::shared_ptr<std::vector<Invalidate>>;
+        InvHandle const vecinv = std::atomic_load(&(pstate->invalidate_vec));
+
+        if (vecinv) {
+          time_t const date    = get_date_from_cached_hdr(txn);
+          time_t const timenow = time(nullptr);
+          char *url            = nullptr;
+          int url_len          = 0;
+
+          for (Invalidate const &inv : *vecinv) {
+            if (date <= inv.epoch && timenow <= inv.expiry) {
+              if (nullptr == url) {
+                url = TSHttpTxnEffectiveUrlStringGet(txn, &url_len);
+              }
+
+              if (nullptr == url || 0 == url_len) {
+                break;
+              } else if (inv.matches(url, url_len)) {
+                TSHttpTxnCacheLookupStatusSet(txn, TS_CACHE_LOOKUP_HIT_STALE);
+                DEBUG_LOG("Forced revalidate - %.*s", url_len, url);
+                break;
+              }
+            }
+          }
+
+          if (nullptr != url) {
+            TSfree(url);
+          }
+        }
+      }
+    }
+    break;
+  case TS_EVENT_HTTP_TXN_CLOSE: // never called by global
+    TSContDestroy(cont);
```

Comment:
I don't see the point of this.  There's no call to put anything on the TXN_CLOSE hook.  main_handler is called by a global continuation, not one that's created exclusively for one transaction.

Reply:
removed.